daemon/c8d: Implement `docker tag`

vvoland opened the pull request:
- Upstreams: https://github.com/rumpl/moby/pull/13


**- What I did**
Implement `docker tag` for containerd image store.

**- How I did it**
See individual commits.

**- How to verify it**
```bash
$ docker pull ubuntu
$ docker tag ubuntu test
$ docker pull alpine
$ docker tag alpine test
```
<details>
<summary>Tag replace</summary>

```bash
$ docker pull alpine
Using default tag: latest
f271e74b17ce: Download complete
41d876d4e443: Download complete
04eeaa5f8c35: Download complete
a9eaa45ef418: Download complete
docker.io/library/alpine:latest
$ docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED                  SIZE
alpine       latest    f271e74b17ce   Less than a second ago   3.26MB
$ docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED         SIZE
alpine       latest    f271e74b17ce   6 seconds ago   3.26MB
$ docker pull busybox
Using default tag: latest
7b3ccabffc97: Download complete
5e42fbc46b17: Download complete
abaa813f94fd: Download complete
f78e6840ded1: Download complete
docker.io/library/busybox:latest
$ docker tag busybox myimage
$ docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED         SIZE
alpine       latest    f271e74b17ce   8 seconds ago   3.26MB
busybox      latest    7b3ccabffc97   5 seconds ago   2.01MB
myimage      latest    7b3ccabffc97   1 second ago    2.01MB
$ docker rmi busybox
Untagged: busybox
$ docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED          SIZE
alpine       latest    f271e74b17ce   13 seconds ago   3.26MB
myimage      latest    7b3ccabffc97   6 seconds ago    2.01MB
$ docker tag alpine myimage
$ docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED          SIZE
<none>       <none>    7b3ccabffc97   2 seconds ago    2.01MB
alpine       latest    f271e74b17ce   24 seconds ago   3.26MB
myimage      latest    f271e74b17ce   2 seconds ago    3.26MB
$ docker tag 7b3ccabffc97 busybox # this needs https://github.com/moby/moby/pull/44842
$ docker image ls
REPOSITORY   TAG       IMAGE ID       CREATED              SIZE
alpine       latest    f271e74b17ce   About a minute ago   3.26MB
busybox      latest    7b3ccabffc97   2 seconds ago        2.01MB
myimage      latest    f271e74b17ce   52 seconds ago       3.26MB


```

</details>


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://user-images.githubusercontent.com/5046555/211568561-d168ab13-bdc0-4a0a-8a78-fef3d9cca2b7.png)

ndeloof committed:
daemon: Pass ctx to image tagging operations

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
daemon/imageService: Remove TagImageWithReference

TagImage is just a wrapper for TagImageWithReference which parses the
repo and tag into a reference. Change TagImageWithReference into
TagImage and move the responsibility of reference parsing to caller.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

rumpl commented on daemon/containerd/image_tag.go:
```diff
@@ -1,20 +1,57 @@
+		err := is.Delete(ctx, img.Name)
+
+		if err != nil {
+			logger.WithError(err).Error("failed to delete old image")
```
The other logs are `Debug` these could be too no?

vvoland replied:
My reasoning was that the first one can be Debug, because it mostly should be just an user error (non existing id) and it's kind of expected to happen normally.

But if the image couldn't be created and it's not an "already exists" error, then probably it's something that's not expected to happen under normal circumstances s and it's worth having logging without `debug` enabled.

thaJeztah replied:
> But if the image couldn't be created and it's not an "already exists" error, then probably it's something that's not expected to happen under normal circumstances s and it's worth having logging without debug enabled.

That one should probably be an `Error()`, as it's an unexpected error (and we currently return a `errdefs.System()` (internal server error).

cpuguy83 replied:
Really we shouldn't need to log and return an error.

vvoland replied:
The log has additional information about the context (imageID, tag) - which might be useful when debugging failures. The only caveat is that these will only appear in the daemon log.
Should we drop the logging and just include these in the returned error?

cpuguy83 replied:
I think it makes sense to include it in the error.

vvoland replied:
Removed the extra `logrus` logs and included details directly in returned errors.

vvoland commented:
CI failures are related, will fix those.

rumpl commented on daemon/containerd/image_tag.go:
```diff
@@ -1,20 +1,57 @@
+		}
+
+		// If there already exists an image with this tag, delete it
+		err := is.Delete(ctx, img.Name)
```
See [this](https://github.com/rumpl/moby/pull/127#discussion_r1070650316) comment, instead of deleting the existing image we should create a dangling image

vvoland replied:
Done

thaJeztah commented on daemon/containerd/image_tag.go:
```diff
@@ -1,20 +1,57 @@
+	if err != nil {
+		if !cerrdefs.IsAlreadyExists(err) {
+			logger.WithError(err).Debug("failed to create image")
+			return errdefs.System(err)
```
Do we know what errors this can return, and if they should always be considered an internal server error? (does `is.Create()` return typed errors from containerd, and can any of those be a "user error" (4xx)?

(Perhaps a 5xx is okay in all cases, but might be worth checking and/or a comment to outline that irregardless of the error type, we consider it an internal server error).

vvoland replied:
Yeah definitely there are some errors which shouldn't be considered a server error, like: missing name, labels too big. Those are the minority of all errors that could be returned though. The rest look system related (grpc, internal storage failures).
I think we should keep this as `System`, but validate the image ourselves before calling `is.Create()` and return an user error.

vvoland replied:
Actually, in this case already have a valid reference which is a valid image name. So it's not possible for this one to fail with user user. Will change the `Debug` to `Error` log then.

rumpl commented on daemon/containerd/image_tag.go:
```diff
@@ -1,20 +1,63 @@
+		}
+
+		// If there already exists an image with this tag, delete it
+		err = i.softImageDelete(oldImg)
```
Nit, `if err := i.softImageDelete(oldImg); err != nil { ... }`

vvoland replied:
Done

vvoland commented on daemon/containerd/soft_delete.go:
```diff
@@ -0,0 +1,77 @@
+		danglingImage := img
+
+		refWithoutTag := reference.TrimNamed(ref)
+		newRef, err := reference.WithDigest(refWithoutTag, img.Target.Digest)
```
This will just discard the tag and still use the original repo.
So for example replacing `docker.io/library/busybox:latest` would change it to `docker.io/library/busybox@sha256:HASH`.
Wondering if we should drop the repo here.
WDYT?

vvoland replied:
Looks like the current Docker drops the repo completely:
```bash
$ docker image ls | grep d3156fec8bcb
alpine-test                        latest                      d3156fec8bcb   8 weeks ago     7.46MB
$ docker tag alpine:3.17.1 alpine-test
$ docker image ls | grep d3156fec8bcb
<none>                             <none>                      d3156fec8bcb   8 weeks ago     7.46MB
$ docker inspect d3156fec8bcb
[
    {
        "Id": "sha256:d3156fec8bcbc7b491a4edc271a7734dcfa186fc73282d4e120eeaaf2ce95c43",
        "RepoTags": [],
        "RepoDigests": [],
...
```

vvoland replied:
Changed, dangling image no longer references the old repository.

vvoland commented on daemon/containerd/image_tag.go:
```diff
@@ -1,20 +1,60 @@
+			return errdefs.System(err)
+		}
+
+		if _, err = is.Create(ctx, img); err != nil {
```
Possibly it would be a good idea to explicitly ignore the user context here and pass `context.Background` instead?
Otherwise we may end up in a situation in which the new tag was deleted and wasn't recreated.

thaJeztah commented on api/server/router/image/backend.go:
```diff
@@ -26,7 +26,7 @@ type imageBackend interface {
 	Images(ctx context.Context, opts types.ImageListOptions) ([]*types.ImageSummary, error)
 	GetImage(ctx context.Context, refOrID string, options image.GetImageOpts) (*dockerimage.Image, error)
-	TagImage(ctx context.Context, imageName, repository, tag string) (string, error)
+	TagImageWithReference(ctx context.Context, id dockerimage.ID, newRef reference.Named) error
```
🚲 🏠 Should we just rename `TagImageWithReference()` to `TagImage()`. I feel like `TagImageWithReference` was named like that because there already was a `TagImage()` but now that that's gone, we may as well use the more "to the point" name.

neersighted replied:
:+1: on this

thaJeztah commented on daemon/containerd/soft_delete.go:
```diff
@@ -0,0 +1,72 @@
+	return nil
+}
+
+func isDanglingImage(image containerdimages.Image) bool {
+	return image.Name == danglingImageName(image.Target.Digest)
+}
```
Wondering if `hasPrefix("dangling@")` would be enough for these 

edit: actually; it looks like `isDanglingImage()` is not used anywhere; was it meant to be used anywhere?

vvoland replied:
It would be enough, but it would also match images which are unlikely to be created by us. User could fiddle with the containerd namespace him self and create an image with name `dangling@home` - which isn't a valid reference, and is not a "dangling" image which we understand.
The whole "dangling" image is our convention, so IMO it's better no to be too liberal in what we consider to be a dangling image.

Yeah it was used first for this branch check: https://github.com/moby/moby/blob/eaf7acbc2951dc40960ceb2e6a436a4494085a07/daemon/containerd/image_tag.go#L69-L80

but then I changed it and it's not used, but it since this is the first upstream PR that creates dangling images I kept it because it will be used by other things (could be used by https://github.com/moby/moby/pull/44840#discussion_r1084202281)

thaJeztah commented on daemon/containerd/image_tag.go:
```diff
@@ -60,7 +60,22 @@ func (i *ImageService) TagImageWithReference(ctx context.Context, imageID image.
-
 	logger.Info("image created")
+
+	if err == nil {
```
Looks like `err != nil` is already handled at line 34 above;

```go
	_, err = is.Create(ctx, newImg)
	if err != nil {
```

vvoland replied:
Good catch, thanks!

thaJeztah commented on daemon/containerd/image_tag.go:
```diff
@@ -12,7 +12,7 @@ import (
 
 // TagImageWithReference creates an image named as newTag and targeting the given descriptor id.
-func (i *ImageService) TagImageWithReference(ctx context.Context, imageID image.ID, newTag reference.Named) error {
+func (i *ImageService) TagImageWithReference(ctx context.Context, imageID image.ID, newTag reference.Named) (retErr error) {
```
I don't see any special handling (defer, e.g.) for the returned error, so we probably don't need a named `retErr` here.

vvoland replied:
Ah good catch, I used it at first for some defer block, but then removed it and forgot to remove the named return. Fixed, thanks!

thaJeztah commented on daemon/containerd/image_tag.go:
```diff
@@ -60,7 +60,22 @@ func (i *ImageService) TagImageWithReference(ctx context.Context, imageID image.
+		// The tag succeeded, check if the source image is dangling
+		sourceDanglingImg, err := is.Get(ctx, danglingImageName(target.Digest))
+
+		if err == nil {
```
Also wondering if it'd be more idiomatic to do an early return instead;

```go
	// The tag succeeded, check if the source image is dangling
	sourceDanglingImg, err := is.Get(ctx, danglingImageName(target.Digest))
	if err != nil {
		if !cerrdefs.IsNotFound(err) {
			logger.WithError(err).Warn("unexpected error when checking if source image is dangling")
		}
		return nil
	}

	// Delete the source dangling image, as it's no longer dangling.
	if err := is.Delete(ctx, sourceDanglingImg.Name); err != nil {
		logger.WithError(err).Warn("unexpected error when deleting dangling image")
	}

	return nil
}
```

vvoland replied:
Changed, thanks

thaJeztah commented on daemon/containerd/image_tag.go:
```diff
@@ -50,11 +50,14 @@ func (i *ImageService) TagImageWithReference(ctx context.Context, imageID image.
 			return errdefs.System(err)
 		}
 
+		// From this point ignore context cancellation.
+		ctx = context.Background()
```
The slightly confusing bit here is that;

- If we didn't get an error on the first `img.Create()`, we get in this "if" branch, and replace the `ctx`  with new context
- However, the `sourceDanglingImg, err := is.Get(ctx, danglingImageName(target.Digest))` also uses the context, so in one case it uses the original context, but in another case it uses a new `context.Background()`.


Should the second (`sourceDanglingImg ... `) code have a new context depending on this? If that also should always be executed with a separate context, perhaps the `context.Background()` should be inlined in both cases (no context shared between the two).

Do any of those contexts need a timeout, or not needed?

vvoland replied:
Right, the code starting from `sourceDanglingImg...` shouldn't depend on the inner code from the previous if branch, but still, if we want to have this "uninterruptible" behavior, then it should independently use other non-cancellable context.

An independent timeout could be a good idea indeed. Just in case there's some unexpected error (grpc/containerd issues?).

tianon commented on daemon/containerd/soft_delete.go:
```diff
@@ -0,0 +1,73 @@
+}
+
+func danglingImageName(digest digest.Digest) string {
+	return "dangling@" + digest.String()
```
Should we make this something like `moby-dangling@xxx` so it's obvious even in `ctr` what this is? :innocent:

vvoland replied:
Sounds good!

vvoland committed:
api/tag: Reject digested target references

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on api/server/router/image/image_routes.go:
```diff
@@ -369,7 +369,29 @@ func (ir *imageRouter) postImagesTag(ctx context.Context, w http.ResponseWriter,
+	repo := r.Form.Get("repo")
+	tag := r.Form.Get("tag")
+
+	ref, err := reference.ParseNormalizedNamed(repo)
+	if err != nil {
+		return errdefs.InvalidParameter(err)
+	}
+
+	if tag != "" {
+		if ref, err = reference.WithTag(reference.TrimNamed(ref), tag); err != nil {
+			return errdefs.InvalidParameter(err)
+		}
+	} else {
+		ref = reference.TagNameOnly(ref)
+	}
```
Yes, this is already checked on the client side (at least in moby client) when splitting reference to repo and tag, but we definitely should also check this on backend side too.
Thanks!

vvoland committed:
c8d/import: Use danglingImageName instead of hardcoded

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

cpuguy83 commented on daemon/containerd/soft_delete.go:
```diff
@@ -0,0 +1,73 @@
+		// Error out in case we couldn't persist the old image.
+		// If it already exists, then just continue.
+		if err != nil && !cerrdefs.IsAlreadyExists(err) {
+			logger.WithError(err).Error("failed to create a dangling image for the image being replaced")
```
Same as above, we shouldn't need to log and return an error.

neersighted reviewed (changes requested):
Would like to see the following changes:

* Amending commits squashed, TODOs between commits minimized
* Don't override `ctx`
* `TagImage` instead of `TagImageWithReference` would be nice

I feel like you may have preferred `TagImageWithReference` over `TagImage` as it makes the need to create the correct reference first clear. It looks like there's common that does this shared between the tag API and the commit code.

Can we consider DRYing that up? It seems possible we could preserve `TagImage` as a fully encapsulated version, though on the other hand, the need for `TagImageWithReference` would be decreased. In any case, I'd like the logic of creating the reference to be de-duplicated and contained in the tagging code, if possible.

vvoland committed:
api: Extract parsing reference from repo and tag

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on daemon/containerd/image_tag.go:
```diff
@@ -1,20 +1,90 @@
+		}
+	}
+
+	logger.Info("image created")
```
It already includes `imageID` and `tag`:
https://github.com/moby/moby/blob/0e41b778680685ed9e59c052fe6aa80f9b59b344/daemon/containerd/image_tag.go#L17-L20

vvoland commented on daemon/containerd/image_tag.go:
```diff
@@ -1,20 +1,90 @@
+	}
+
+	// From this point ignore context cancellation, but set a realistic timeout just in case
+	freshCtx, release := context.WithTimeout(context.Background(), time.Second*15)
```
Removed this one in favor of plain `context.Background`.

vvoland reviewed (commented):
Squashed, removed the ctx overriding and changed `TagImageWithReference` into `TagImage`.

I still kept the reference parsing logic outside of tagging code - I don't see how it belongs here, many operations should make sure the reference is valid, before attempting the tag, so they already have a parsed reference. Also it feel weird to make `ImageService` responsible for parsing that.
I moved the reference parsing from commit to the caller code and made it accept a `NamedParsed` reference instead. The parsing logic is extracted to httputils as for now only http handlers parse the reference, but ideally this should probably belong to the `reference` package,

neersighted reviewed (changes requested):
Updating my review: LGTM with the exception of dropping the unused method causing a linter error.

vvoland committed:
daemon/c8d: Implement TagImageWithReference

Implements image tagging under containerd image store.

If an image with this tag already exists, and there's no other image
with the same target, change its name. The name will have a special
format `moby-dangling@<digest>` which isn't a valid canonical reference
and doesn't resolve to any repository.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

neersighted commented:
I'd like to see @cpuguy83's logging concerns addressed (looks like he didn't mark them as changes requested). Looks ready to merge once he gives his sign-off.

neersighted commented:
Let's :shipit:

rumpl reviewed (approved):
👍👍👍

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
arf.. running into the "dubious permissions" issue; @vvoland can you try rebasing?

vvoland commented:
@thaJeztah Rebased

thaJeztah commented:
Thanks!
